### PR TITLE
Add human readable error message for invalid main signatures

### PR DESCRIPTION
**Pull request by hannes-steffenhagen-diffblue**
Currently we have special handling if the entry point for analysis is called
`main`, in particular we assume it is one of the following:

    int main(void)
    int main(int argc, char *argv[])
    int main(int argc, char *argv[], char *envp[])

If it is not so far we have been failing with an UNREACHABLE invariant failure.
Since it is completely possible to have a main signature like that, we should
fail with a proper human readable error message that tells people what is wrong
and maybe how they can work around it.

For now, addresses https://github.com/diffblue/cbmc/issues/5415. As I say in that ticket maybe we want to handle non-standard `main` signatures in a more graceful way, but until we decide on that having a better error message should take away some of the pain.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by NlightNFotis — APPROVED** [review]
Okay

**Comment by NlightNFotis** [comment]
Is there a long term plan for better handling of the situation? Or is it how we're going to work with this?

**Comment by hannes-steffenhagen-diffblue** [comment]
@NlightNFotis Like I said, I am not sure what the ‘better’ version looks like.

File: src/ansi-c/ansi_c_entry_point.cpp [hunk 1]
```diff
@@ -473,7 +474,19 @@ bool generate_ansi_c_start_function(
...
+      const std::string main_signature = type2c(symbol.type, ns);
+      throw invalid_source_file_exceptiont{
+        "'main' with signature '" + main_signature +
+        "' found,"
+        " but expecting one of:\n"
+        "   int main(void)\n"
+        "   int main(int argc, char *argv[])\n"
+        "   int main(int argc, char *argv[], char *envp[])\n"
+        "If this is a non-standard main entry point please provide a custom\n"
+        "entry function and point to it via cbmc --function instead"};
```
Comment (thk123) [inline src/ansi-c/ansi_c_entry_point.cpp:488]: This is a great error message :green_heart: 
:grey_question: If I do `cbmc main.c --function main` does this work? No worries if not, but if it does, I'd be half tempted to give that as the advice

**Review by thk123 — APPROVED** [review]

File: src/ansi-c/ansi_c_entry_point.cpp [hunk 2]
```diff
@@ -473,7 +474,19 @@ bool generate_ansi_c_start_function(
       }
     }
     else
-      UNREACHABLE;
+    {
+      const namespacet ns{symbol_table};
+      const std::string main_signature = type2c(symbol.type, ns);
+      throw invalid_source_file_exceptiont{
+        "'main' with signature '" + main_signature +
```
Comment (chrisr-diffblue) [inline src/ansi-c/ansi_c_entry_point.cpp:481]: Should it explicitly say "with an invalid signature" ?

File: src/ansi-c/ansi_c_entry_point.cpp [hunk 3]
```diff
@@ -473,7 +474,19 @@ bool generate_ansi_c_start_function(
...
+        "' found,"
+        " but expecting one of:\n"
+        "   int main(void)\n"
+        "   int main(int argc, char *argv[])\n"
+        "   int main(int argc, char *argv[], char *envp[])\n"
+        "If this is a non-standard main entry point please provide a custom\n"
```
Comment (chrisr-diffblue) [inline src/ansi-c/ansi_c_entry_point.cpp:487]: Thanks for the extra user help in the message :-)

**Review by chrisr-diffblue — APPROVED** [review]

File: src/ansi-c/ansi_c_entry_point.cpp (hunk 1 above)
Comment (hannes-steffenhagen-diffblue) [reply src/ansi-c/ansi_c_entry_point.cpp:488]: It doesn’t unfortunately. What does work is adding something like

```
void harness(void) {
  main(NULL);
}
```

and pointing at that.

**Review by thomasspriggs — APPROVED** [review]
:+1:

File: src/ansi-c/ansi_c_entry_point.cpp (hunk 2 above)
Comment (hannes-steffenhagen-diffblue) [reply src/ansi-c/ansi_c_entry_point.cpp:481]: Well, I feel like ‘invalid’ is a matter of opinion. Maybe this is coming from code that does have a non-standard entry point. It is certainly not what we expect though, so saying that much is value neutral.

**Pull request merged**